Add first & last_id segments to collection cache_version

alipman88 opened the pull request:
### Summary

In collections with a limit, including the IDs of the first and last records when generating a collection's cache version ensures correct cache key expiration after a record is deleted. Prior to this commit, collection cache versions only include the collection's size and maximum updated_at timestamp, e.g.:

    "#{collection_size}-#{max_updated_at_timestamp}"

To demonstrate why this isn't robust enough, consider the following database table:

    Employee.all

    SELECT * FROM employees;
    +----+-------+---------------------+---------------------+
    | id | name  | created_at          | updated_at          |
    +----+-------+---------------------+---------------------+
    |  1 | alice | 2019-11-01 00:00:00 | 2019-11-01 00:00:00 |
    |  2 | carol | 2019-11-02 00:00:00 | 2019-11-02 00:00:00 |
    |  3 | doug  | 2019-11-03 00:00:00 | 2019-11-03 00:00:00 |
    |  4 | bob   | 2019-11-04 00:00:00 | 2019-11-04 00:00:00 |
    +----+-------+---------------------+---------------------+

Select the first three employees sorted alphabetically:

    employees = Employee.order(:name).limit(3)

    SELECT * FROM employees ORDER BY name ASC LIMIT 3;
    +----+-------+---------------------+---------------------+
    | id | name  | created_at          | updated_at          |
    +----+-------+---------------------+---------------------+
    |  1 | alice | 2019-11-01 00:00:00 | 2019-11-01 00:00:00 |
    |  4 | bob   | 2019-11-04 00:00:00 | 2019-11-04 00:00:00 |
    |  2 | carol | 2019-11-02 00:00:00 | 2019-11-02 00:00:00 |
    +----+-------+---------------------+---------------------+

Note the collection's size (3) and max updated_at timestamp (2019-11-04 00:00:00), both of which are used to compose the cache version:

    employees.cache_version
    # => "3-20191104000000"

Now, delete an employee:

    Employee.find_by(name: "alice").destroy

    DELETE FROM employees WHERE name = "alice";

Re-run the original query:

    employees = Employee.order(:name).limit(3)

    SELECT * FROM employees ORDER BY name ASC LIMIT 3;
    +----+-------+---------------------+---------------------+
    | id | name  | created_at          | updated_at          |
    +----+-------+---------------------+---------------------+
    |  4 | bob   | 2019-11-04 00:00:00 | 2019-11-04 00:00:00 |
    |  2 | carol | 2019-11-02 00:00:00 | 2019-11-02 00:00:00 |
    |  3 | doug  | 2019-11-03 00:00:00 | 2019-11-03 00:00:00 |
    +----+-------+---------------------+---------------------+

Note that while the collection's contents have changed, its size and maximum updated_at timestamp remain the same, resulting in the cache_version failing to correctly expire:

    employees.cache_version
    # => "3-20191104000000"

Including the ID of the first and last record in a collection fixes this problem - when deleting a record from a collection with a limit (or some other combination of inserts/updates/deletes), either the collection size will shrink, its max updated_at timestamp will increase, or one of its bounding IDs will change.

If a collection is loaded, determining the IDs of its first and last records is straightforward:

    employees = Employee.limit(3)
    first_id = employees.size > 0 ? employees.first.id : nil
    last_id = employees.size > 0 ? employees.last.id : nil

If a collection isn't loaded, this commit uses SQL common table expressions (for SQLite and PostgreSQL databases) or user-defined variables (for MySQL databases) to determine the last record's ID. Example SQL queries below (table aliases shortened for readability):

    -- PostgreSQL/SQLite
    WITH
      col AS
        (SELECT id, updated_at FROM employees ORDER BY name ASC LIMIT 3),
      agg AS
        (SELECT COUNT(*) AS size, MAX(updated_at) AS timestamp FROM col)
    SELECT
      size,
      timestamp,
      (SELECT id FROM col LIMIT 1) AS first_id,
      (SELECT id FROM col LIMIT 1 OFFSET
        GREATEST((SELECT size FROM agg) - 1, 0)) AS last_id
    FROM agg;

    -- MySQL
    SET @first_id := NULL, @last_id := NULL;
    SELECT
      COUNT(*) AS collection_size,
      MAX(updated_at) AS timestamp,
      @first_id AS first_id,
      @last_id AS last_id
    FROM (
      SELECT
        @first_id := COALESCE(@first_id, id),
        @last_id := id,
        updated_at
      FROM employees
      ORDER BY name ASC LIMIT 3
    ) col;

    +------+---------------------+----------+---------+
    | size | timestamp           | first_id | last_id |
    +------+---------------------+----------+---------+
    |    3 | 2019-11-04 00:00:00 |        4 |       3 |
    +------+---------------------+----------+---------+

Fixes #31996, #34408 and #37555.

### Benchmarks

This change has a negligible impact on common use-cases (e.g. paginating records). For very large collections (~100k+ records for Postgres, ~10k+ for MySQL), generating cache_versions takes an estimated 1.75 times longer for Postgres, and 2.75 for MySQL. However, this change should not impact the overall time complexity which should remain linear in regard to the limit/collection size.

Here are some charts (with linear and logarithmic x axes) visualizing the results of my benchmarking (full script and raw results further below):

https://docs.google.com/spreadsheets/d/1d9u7vPh7AD8ArD4krRY7esWsvoBP1eK925prOH4c1uU/edit?folder=0AL-x-3mNk-iPUk9PVA#gid=1452480752

Log charts:

![postgres log chart](https://docs.google.com/spreadsheets/d/e/2PACX-1vTvSS1smbMOibaMpOLHr1ryO6aun0VZzTsJsDe4M2gWGdWZ6BywGZtGvDA_neH1-7rhK77A9NqQS9WB/pubchart?oid=1118001582&format=image)

![mysql log chart](https://docs.google.com/spreadsheets/d/e/2PACX-1vTvSS1smbMOibaMpOLHr1ryO6aun0VZzTsJsDe4M2gWGdWZ6BywGZtGvDA_neH1-7rhK77A9NqQS9WB/pubchart?oid=345571619&format=image)

Linear charts:

![postgres log chart](https://docs.google.com/spreadsheets/d/e/2PACX-1vTvSS1smbMOibaMpOLHr1ryO6aun0VZzTsJsDe4M2gWGdWZ6BywGZtGvDA_neH1-7rhK77A9NqQS9WB/pubchart?oid=133526580&format=image)

![mysql log chart](https://docs.google.com/spreadsheets/d/e/2PACX-1vTvSS1smbMOibaMpOLHr1ryO6aun0VZzTsJsDe4M2gWGdWZ6BywGZtGvDA_neH1-7rhK77A9NqQS9WB/pubchart?oid=1389627901&format=image)

### Other changes

This commit updates the signatures of `ActiveRecord::Relation#cache_version`, `#cache_key` and related methods to use keyword instead of positional arguments, as said methods now accept multiple optional arguments. In order to avoid breaking changes, both methods still accept positional arguments, but will raise deprecation warnings if passed.

### Other considered solutions

In writing this commit, I reviewed other potential solutions to this issue, particularly https://github.com/rails/rails/pull/21503. Most other solutions detected changes to a collection via either a checksum of its IDs, or a hash digest of a string of concatenated IDs. I leaned away from this approach for several reasons:

1. SQLite doesn't have an MD5 function.
2. It's inefficient for very large collections, taking about four times longer than the current approach. (Although the overall time complexity appears equivalent.)
3. For HABTM relations where neither model's updated_at columns are touched when associations are created or destroyed, an ID checksum covers one additional use case: On a query like `@project.employees`, an ID checksum would detect changes to a project's employees even when the collection size, max updated_at timestamp and bounding IDs remain the same. However, an ID checksum *won't* cover the most basic use case of Russian doll caching: consider a cached collection of projects each with an inner-nested cached collection of employees - changes to a project's employees fails to expire the outer cached fragment. Thus, I don't find that the ID checksum approach provides a comprehensive advantage.

### Benchmarking Script + results

For the purposes of Benchmarking, I provisioned Postgres and MySQL databases through Heroku's add-ons. For Postgres, I used a "Standard-0" tier DB with 4 GB ram. For MySQL, I used a JawsSB "Whitetip" tier DB with 1 GB of ram. (Both services are backed by AWS.)

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails"
  gem "benchmark-ips"
  gem "pg"
  gem "mysql2"
end

require "active_record"
require "logger"

ActiveRecord::Base.establish_connection({
  # redacted
})

ActiveRecord::Base.collection_cache_versioning = true

class ActiveRecord::Relation
  def cache_version_new(deprecated_timestamp_column = nil, id_column: primary_key, timestamp_column: :updated_at)
    if deprecated_timestamp_column
      ActiveSupport::Deprecation.warn("Passing a timestamp column as a positional argument to ActiveRecord::Relation#cache_version is deprecated. Please use a `:timestamp_column` keyword argument instead.")
      timestamp_column = deprecated_timestamp_column
    end

    if collection_cache_versioning
      @cache_versions_new ||= {}
      @cache_versions_new[[id_column, timestamp_column]] ||= compute_cache_version_new(id_column: id_column, timestamp_column: timestamp_column)
    end
  end

  def compute_cache_version_new(id_column: primary_key, timestamp_column: :updated_at) # :nodoc:
    size = 0
    first_id = nil
    last_id = nil
    timestamp = nil

    if loaded? || distinct_value
      size = records.size

      if size > 0
        timestamp = max_by(&timestamp_column)._read_attribute(timestamp_column).utc.to_s(cache_timestamp_format)

        if has_limit_or_offset?
          first_id = first._read_attribute(id_column)
          last_id = last._read_attribute(id_column)
        end
      end
    else
      collection = eager_loading? ? apply_join_dependency : self
      timestamp_col = connection.visitor.compile(arel_attribute(timestamp_column))
      id_col = connection.visitor.compile(arel_attribute(id_column))

      if collection.has_limit_or_offset?
        case connection.adapter_name
        when "PostgreSQL", "SQLite"
          collection_table = Arel::Table.new("cache_key_collection")
          aggregates_table = Arel::Table.new("cache_key_aggregates")
          query = collection.select("#{id_col} AS collection_cache_key_id, #{timestamp_col} AS collection_cache_key_timestamp")
          collection_query = Arel::Nodes::SqlLiteral.new "(#{ query.to_sql })"
          aggregates_query = collection_table.project("COUNT(*) AS collection_size, MAX(collection_cache_key_timestamp) AS timestamp")
          collection_common_table_expr = Arel::Nodes::As.new(collection_table, collection_query)
          aggregates_common_table_expr = Arel::Nodes::As.new(aggregates_table, aggregates_query)

          collection_size_minus_one = Arel::Nodes::Subtraction.new(aggregates_table[:collection_size], 1)
          last_id_offset = connection.adapter_name == "PostgreSQL" ?
            Arel::Nodes::NamedFunction.new("GREATEST", [collection_size_minus_one, 0]) :
            aggregates_table.project(collection_size_minus_one)

          select_values = [
            aggregates_table[:collection_size],
            aggregates_table[:timestamp],
            collection_table.project(collection_table[:collection_cache_key_id]).take(1).as("first_id"),
            collection_table.project(collection_table[:collection_cache_key_id]).take(1).skip(last_id_offset).as("last_id")
          ]

          arel = aggregates_table
            .project(select_values)
            .with(collection_common_table_expr, aggregates_common_table_expr)
        when "Mysql2"
          select_values = "COUNT(*) AS collection_size, @first_id AS first_id, @last_id AS last_id, MAX(collection_cache_key_timestamp) AS timestamp"
          query = collection.select("COALESCE(@first_id, @first_id := #{id_col}), @last_id := #{id_col}, #{timestamp_col} AS collection_cache_key_timestamp")
          subquery_alias = "subquery_for_cache_key"
          arel = query.build_subquery(subquery_alias, select_values)
        else
          select_values = "COUNT(*) AS collection_size, MAX(#{timestamp_col}) AS timestamp"
          query = collection.select("#{timestamp_col} AS collection_cache_key_timestamp")
          subquery_alias = "subquery_for_cache_key"
          arel = query.build_subquery(subquery_alias, select_values)
        end
      else
        select_values = "COUNT(*) AS collection_size, MAX(#{timestamp_col}) AS timestamp"
        query = collection.unscope(:order)
        query.select_values = [select_values]
        arel = query.arel
      end

      result = nil

      connection_pool.with_connection do
        if connection.adapter_name == "Mysql2" && collection.has_limit_or_offset?
          connection.execute("SET @first_id := @last_id := NULL")
        end

        result = connection.select_one(arel, nil)
      end

      if result
        column_type = klass.type_for_attribute(timestamp_column)
        size = result["collection_size"]

        if size > 0
          first_id = result["first_id"]
          last_id = result["last_id"]
          timestamp = column_type.deserialize(result["timestamp"]).utc.to_s(cache_timestamp_format)
        end
      end
    end

    [size, first_id, last_id, timestamp].reject(&:blank?).join("-")
  end
  private :compute_cache_version_new
end

ActiveRecord::Schema.define do
  create_table :users, if_not_exists: true do |t|
    t.string :name
    t.timestamps
  end
end

class User < ActiveRecord::Base
end

# Seed database with two million records
if User.first.nil?
  r = 2_000_000
  t = 5.years.ago.to_i
  now = Time.now.to_i
  pct = 0.0

  puts "initializing test records"

  (1..r).map { |n| {
    id: n,
    name: SecureRandom.alphanumeric(5 + rand(12)),
    created_at: Time.at(t += (now - t) / (1 + r - n)),
    updated_at: Time.at(rand(t..now+1))
  } }.each_slice(20_000) { |slice|
    puts "seeding DB" if pct == 0.0
    User.insert_all(slice)
    puts "#{pct+=1}%"
  }
end

Benchmark.ips do |benchmark|
  benchmark.config(warmup: 0, time: 120)

  limits = [10, 100, 1000, 10000, 100000, 200000, 400000, 600000, 800000, 1000000]
  orders = [:asc, :desc]
  offsets = (0..1_000_000).step(250_000)

  limits.each do |limit|
    benchmark.report("current cache_version, limit #{limit}") do
      orders.each do |order|
        offsets.each do |offset|
          User.order(id: order).limit(limit).offset(offset).cache_version
        end
      end
    end

    benchmark.report("w/ first & last ID, limit #{limit}") do
      orders.each do |order|
        offsets.each do |offset|
          User.order(id: order).limit(limit).offset(offset).cache_version_new
        end
      end
    end
  end
end
```

results:

```
Postgres
Calculating -------------------------------------
current cache_version, limit 10
                          0.457  (± 0.0%) i/s -     53.000  in 120.940372s
w/ first & last ID, limit 10
                          0.630  (± 0.0%) i/s -     74.000  in 120.222423s
current cache_version, limit 100
                          0.555  (± 0.0%) i/s -     66.000  in 121.112922s
w/ first & last ID, limit 100
                          0.622  (± 0.0%) i/s -     71.000  in 120.100475s
current cache_version, limit 1000
                          0.473  (± 0.0%) i/s -     55.000  in 120.244607s
w/ first & last ID, limit 1000
                          0.627  (± 0.0%) i/s -     71.000  in 120.659220s
current cache_version, limit 10000
                          0.442  (± 0.0%) i/s -     52.000  in 121.285019s
w/ first & last ID, limit 10000
                          0.561  (± 0.0%) i/s -     64.000  in 120.059518s
current cache_version, limit 100000
                          0.431  (± 0.0%) i/s -     51.000  in 121.234291s
w/ first & last ID, limit 100000
                          0.410  (± 0.0%) i/s -     48.000  in 120.316814s
current cache_version, limit 200000
                          0.406  (± 0.0%) i/s -     49.000  in 121.883819s
w/ first & last ID, limit 200000
                          0.303  (± 0.0%) i/s -     37.000  in 122.960886s
current cache_version, limit 400000
                          0.315  (± 0.0%) i/s -     38.000  in 121.148364s
w/ first & last ID, limit 400000
                          0.223  (± 0.0%) i/s -     27.000  in 121.430635s
current cache_version, limit 600000
                          0.266  (± 0.0%) i/s -     32.000  in 120.808200s
w/ first & last ID, limit 600000
                          0.168  (± 0.0%) i/s -     21.000  in 125.029394s
current cache_version, limit 800000
                          0.236  (± 0.0%) i/s -     29.000  in 123.167375s
w/ first & last ID, limit 800000
                          0.140  (± 0.0%) i/s -     17.000  in 121.361332s
current cache_version, limit 1000000
                          0.208  (± 0.0%) i/s -     26.000  in 125.224578s
w/ first & last ID, limit 1000000
                          0.120  (± 0.0%) i/s -     15.000  in 125.416142s

MySQL
current cache_version, limit 10
                          0.400  (± 0.0%) i/s -     46.000  in 120.437852s
w/ first & last ID, limit 10
                          0.429  (± 0.0%) i/s -     51.000  in 121.507532s
current cache_version, limit 100
                          0.380  (± 0.0%) i/s -     45.000  in 122.325720s
w/ first & last ID, limit 100
                          0.426  (± 0.0%) i/s -     51.000  in 122.081295s
current cache_version, limit 1000
                          0.436  (± 0.0%) i/s -     50.000  in 121.347133s
w/ first & last ID, limit 1000
                          0.422  (± 0.0%) i/s -     50.000  in 120.497030s
current cache_version, limit 10000
                          0.424  (± 0.0%) i/s -     49.000  in 120.461740s
w/ first & last ID, limit 10000
                          0.395  (± 0.0%) i/s -     46.000  in 120.202285s
current cache_version, limit 100000
                          0.304  (± 0.0%) i/s -     35.000  in 120.550654s
w/ first & last ID, limit 100000
                          0.198  (± 0.0%) i/s -     24.000  in 122.127827s
current cache_version, limit 200000
                          0.215  (± 0.0%) i/s -     26.000  in 122.966909s
w/ first & last ID, limit 200000
                          0.150  (± 0.0%) i/s -     18.000  in 120.493927s
current cache_version, limit 400000
                          0.173  (± 0.0%) i/s -     21.000  in 121.901465s
w/ first & last ID, limit 400000
                          0.100  (± 0.0%) i/s -     12.000  in 120.903169s
current cache_version, limit 600000
                          0.156  (± 0.0%) i/s -     19.000  in 122.053916s
w/ first & last ID, limit 600000
                          0.079  (± 0.0%) i/s -     10.000  in 126.504833s
current cache_version, limit 800000
                          0.147  (± 0.0%) i/s -     18.000  in 123.154995s
w/ first & last ID, limit 800000
                          0.062  (± 0.0%) i/s -      8.000  in 128.516317s
current cache_version, limit 1000000
                          0.129  (± 0.0%) i/s -     16.000  in 124.396169s
w/ first & last ID, limit 1000000
                          0.048  (± 0.0%) i/s -      6.000  in 126.030320s
```

alipman88 committed:
Add first & last IDs to collection cache versions

In collections with a limit, including the ID of the first and last
record when generating a collection's cache version ensures correct
cache key expiration after a record is deleted. Prior to this commit,
collection cache versions only include the collection's size and maximum
updated_at timestamp, e.g.:

    "#{collection_size}-#{max_updated_at_timestamp}"

To demonstrate why this isn't robust enough, consider the following
database table:

    Employee.all

    SELECT * FROM employees;
    +----+-------+---------------------+---------------------+
    | id | name  | created_at          | updated_at          |
    +----+-------+---------------------+---------------------+
    |  1 | alice | 2019-11-01 00:00:00 | 2019-11-01 00:00:00 |
    |  2 | carol | 2019-11-02 00:00:00 | 2019-11-02 00:00:00 |
    |  3 | doug  | 2019-11-03 00:00:00 | 2019-11-03 00:00:00 |
    |  4 | bob   | 2019-11-04 00:00:00 | 2019-11-04 00:00:00 |
    +----+-------+---------------------+---------------------+

Select the first three employees sorted alphabetically:

    employees = Employee.order(:name).limit(3)

    SELECT * FROM employees ORDER BY name ASC LIMIT 3;
    +----+-------+---------------------+---------------------+
    | id | name  | created_at          | updated_at          |
    +----+-------+---------------------+---------------------+
    |  1 | alice | 2019-11-01 00:00:00 | 2019-11-01 00:00:00 |
    |  4 | bob   | 2019-11-04 00:00:00 | 2019-11-04 00:00:00 |
    |  2 | carol | 2019-11-02 00:00:00 | 2019-11-02 00:00:00 |
    +----+-------+---------------------+---------------------+

Note the collection's size (3) and max updated_at timestamp
(2019-11-04 00:00:00), both of which are used to compose the cache
version:

    employees.cache_version
    # => "3-20191104000000"

Now, delete an employee:

    Employee.find_by(name: "alice").destroy

    DELETE FROM employees WHERE name = "alice";

Re-run the original query:

    employees = Employee.order(:name).limit(3)

    SELECT * FROM employees ORDER BY name ASC LIMIT 3;
    +----+-------+---------------------+---------------------+
    | id | name  | created_at          | updated_at          |
    +----+-------+---------------------+---------------------+
    |  4 | bob   | 2019-11-04 00:00:00 | 2019-11-04 00:00:00 |
    |  2 | carol | 2019-11-02 00:00:00 | 2019-11-02 00:00:00 |
    |  3 | doug  | 2019-11-03 00:00:00 | 2019-11-03 00:00:00 |
    +----+-------+---------------------+---------------------+

Note that while the contents of the collection has changed, its size and
maximum updated_at timestamps remain the same, resulting in the
cache_version failing to correctly expire:

    employees.cache_version
    # => "3-20191104000000"

Including the ID of the first and last records in a collection fixes
this problem - when deleting a record from a collection with a limit,
either the collection size will shrink or one of its bounding IDs will
change.

If a collection is loaded, determining the IDs of its first and last
record is straightforward:

    employees = Employee.limit(3)
    first_id = employees.size > 0 ? employees.first.id : nil
    last_id = employees.size > 0 ? employees.last.id : nil

If a collection isn't loaded, this commit uses SQL common table
expressions (for SQLite and PostgreSQL databases) or user-defined
variables (for MySQL databases) to determine the collection's first and
last IDs. Example SQL queries below (table and column aliases shortened
for readability):

    -- PostgreSQL/SQLite
    WITH
      col AS
        (SELECT id, updated_at FROM users ORDER BY name ASC LIMIT 3),
      agg AS
        (SELECT COUNT(*) AS size, MAX(updated_at) AS timestamp FROM col)
    SELECT
      size,
      timestamp,
      (SELECT id FROM col LIMIT 1) AS first_id,
      (SELECT id FROM col LIMIT 1 OFFSET
        GREATEST((SELECT size FROM agg) - 1, 0)) AS last_id
    FROM agg;

    -- MySQL
    SET @first_id := NULL, @last_id := NULL;
    SELECT
      COUNT(*) AS size,
      MAX(updated_at) AS timestamp,
      @first_id AS first_id,
      @last_id AS last_id
    FROM (
      SELECT
        @first_id := COALESCE(@first_id, id),
        @last_id := id,
        updated_at
      FROM users
      ORDER BY name ASC LIMIT 3
    ) col;

    +------+----------+---------+---------------------+
    | size | first_id | last_id | timestamp           |
    +------+----------+---------+---------------------+
    |    3 |        4 |       3 | 2019-11-04 00:00:00 |
    +------+----------+---------+---------------------+

Additionally, this commit updates the signatures of
ActiveRecord::Relation#cache_version, #cache_key and related methods to
use keyword instead of positional arguments, as said methods now accept
multiple optional arguments. In order to avoid breaking changes, both
methods still accept positional arguments, but will raise deprecation
warnings if passed.

Fixes #31996, #34408 and #37555.

jonathanhefner commented:
Thank you for attacking this issue!  :pray:  :tada:  Unfortunately, there is a counter-case for this approach which appears in #34408.  In #34408, the last page of results changes from `[7, 8, 9]` to `[6, 7, 9]`, but the cache key would remain `"3-12:09-9"` when using this last ID approach.  However, if you used both first and last ID, the cache key would change appropriately.  So the question is: are there also counter-cases that aren't solved by first and last ID (and count and timestamp)?

It seems like there should be, but I can't think of one.  Here are some scenarios:

* record deleted
  * that record's page => last ID or count changes
  * previous pages => no change necessary
  * subsequent pages => last ID or count changes
* record inserted
  * that record's page => max timestamp changes
  * previous pages => no change necessary
  * subsequent pages => last ID or count changes
* record updated / stayed on same results page
  * that record's page => max timestamp changes
  * previous pages => no change necessary
  * subsequent pages => no change necessary
* record updated / moved upward in result pages
  * that record's new page => max timestamp changes
  * that record's old page => first ID changes
  * pages in between => first and last ID change
  * other pages => no change necessary
* record updated / moved downward in result pages
  * that record's new page => max timestamp changes
  * that record's old page => last ID changes
  * pages in between => first and last ID change
  * other pages => no change necessary

Are all of these correct?  And are there cases where combining two or more scenarios might break the solution?

Also, since performance was a concern for #21503, have you benchmarked this approach?

alipman88 commented:
Thanks for identifying my oversight and proposing a more robust solution, jonathanhefner!

I believe first + last ID does the trick. Consider any page of results:

* If any combination of transactions (inserts/updates/deletes) increases or decreases the number of rows before a page's first ID, the first ID will have shifted when recalculating the cache_version.

* If any combination of transactions changes the number of rows between a page's first ID and last ID, when recalculating the cache version either one or the other will have shifted, or the collection size will have shrunk.

* If any combination of transactions modifies records contained in a page without changing its first ID or last ID, the collection's max updated_at timestamp will have increased.

I was remiss in not including benchmarks originally. Currently, the time required to calculate a cache_version generally appears to grow linearly in regard to the size of a collection. This is because in order to calculate the maximum updated_at column from a subquery of N rows, the database needs to scan through all N rows. (The exception is when the subquery itself isn't efficient, e.g. it's sorted by a non-indexed column - in this case calculating the cache_version will be slow regardless of the algorithm used to do so.)

I believe that in terms of computational complexity/big O notation, both the current cache_key system SQL queries and all the proposed solutions (including the MD5 checksum) run in linear time relative to collection size/limit - so none of them are "horrible."

When including both the first and last ID, I found that using SQL common table expressions slightly improved on my original approach of SQL window functions.

For the purposes of benchmarking, I used a Heroku Postgres DB (standard-0 tier, 4 GB RAM), inserted 2 million records, and with collection sizes/limits ranging from one hundred to one million, benchmarked the time it took for each cache_version algorithm/query to run eight times (four different offsets, ascending & descending sort orders).

The full script and results are at the bottom, but in quick summary: on large collections, the first and last ID approach takes about 50% longer than the current approach, and the MD5 checksum approach takes about 290% longer than the current approach:

https://docs.google.com/spreadsheets/d/1VEH6Rq2lN7nyNCwbVNC5lnOs2TKa1R8Vgh9Vc3_UdcY/edit#gid=742319261

I'm still considering how to elegantly capture the first ID in MySQL - once I feel I've figured that out, I'll update my pull request, add updated benchmarks, and let folks know. (I think it might be best to keep this pull request open so the conversation stays in one place, but the Rails team can feel free to close it if they'd prefer a fresh PR.)

Benchmarking script:

```
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails"
  gem "benchmark-ips"
  gem "pg"
end

require "active_record"
require "logger"

ActiveRecord::Base.collection_cache_versioning = true

ActiveRecord::Base.establish_connection({
  adapter: "postgresql",
  url: "redacted"
})

ActiveRecord::Schema.define do
  create_table :users, if_not_exists: true do |t|
    t.string :name
    t.timestamps
  end
end

class User < ActiveRecord::Base
end

# Seed database with one million records
if User.first.nil?
  r = 2_000_000
  t = 5.years.ago.to_i
  now = Time.now.to_i
  pct = 0.0

  puts "initializing test records"

  (1..r).map { |n| {
    id: n,
    name: SecureRandom.alphanumeric(5 + rand(12)),
    created_at: Time.at(t += (now - t) / (1 + r - n)),
    updated_at: Time.at(rand(t..now+1))
  } }.each_slice(20_000) { |slice|
    puts "seeding BD" if pct == 0.0
    User.insert_all(slice)
    puts "#{pct+=1}%"
  }
  
  records.each_slice(20_000) { |slice| User.insert_all(slice) ; puts "#{pct+=1}%" }
end

Benchmark.ips do |benchmark|
  benchmark.config({warmup: 0, time: 60})

  limits = (0..1_000_000).step(100_000).to_a
  limits [0] = 100
  orders = ["ASC", "DESC"]
  offsets = (0..1_000_000).step(250_000)

  limits.each do |limit|
    benchmark.report("current cache_version, limit #{limit}") do
      orders.each do |order|
        offsets.each do |offset|
          ActiveRecord::Base.connection.select_one %(
            SELECT
              COUNT(*) AS size,
              MAX(updated_at) AS timestamp
            FROM (SELECT updated_at FROM users ORDER BY id #{order} LIMIT #{limit} OFFSET #{offset}) q
          )
        end
      end
    end

    benchmark.report("common table expressions, limit #{limit}") do
      orders.each do |order|
        offsets.each do |offset|
          ActiveRecord::Base.connection.select_one %(
            WITH
              q1 AS (SELECT id, updated_at FROM users ORDER BY id #{order} LIMIT #{limit} OFFSET #{offset}),
              q2 AS (SELECT COUNT(*) AS size, MAX(updated_at) AS timestamp FROM q1)
            SELECT
              q2.size,
              q2.timestamp,
              (SELECT id FROM q1 LIMIT 1) AS fst_id,
              (SELECT id FROM q1 LIMIT 1 OFFSET q2.size - 1) AS lst_id
            FROM q2
          )
        end
      end
    end

    benchmark.report("cache_version w/ MD5 checksum, limit #{limit}") do
      orders.each do |order|
        offsets.each do |offset|
          ActiveRecord::Base.connection.select_one %(
            SELECT
              COUNT(*) AS size,
              MAX(updated_at) AS timestamp,
              SUM(18446744073709551616 * ('x' || LEFT(MD5(id::text), 16))::bit(64)::bigint::decimal(58) +
                ('x' || RIGHT(MD5(id::text), 16))::bit(64)::bigint::decimal(58)) AS checksum
            FROM (SELECT id, updated_at FROM users ORDER BY id #{order} LIMIT #{limit} OFFSET #{offset}) q
          )
        end
      end
    end
  end
end
```

Results:
```
Calculating -------------------------------------
current cache_version, limit 100
                          0.446  (± 0.0%) i/s -     26.000  in  60.715002s
common table expressions, limit 100
                          0.498  (± 0.0%) i/s -     29.000  in  60.378481s
cache_version w/ MD5 checksum, limit 100
                          0.446  (± 0.0%) i/s -     26.000  in  60.295986s
current cache_version, limit 100000
                          0.402  (± 0.0%) i/s -     24.000  in  61.291920s
common table expressions, limit 100000
                          0.358  (± 0.0%) i/s -     22.000  in  62.967460s
cache_version w/ MD5 checksum, limit 100000
                          0.171  (± 0.0%) i/s -     11.000  in  65.586271s
current cache_version, limit 200000
                          0.312  (± 0.0%) i/s -     19.000  in  62.331424s
common table expressions, limit 200000
                          0.255  (± 0.0%) i/s -     15.000  in  60.186287s
cache_version w/ MD5 checksum, limit 200000
                          0.126  (± 0.0%) i/s -      8.000  in  63.850849s
current cache_version, limit 300000
                          0.339  (± 0.0%) i/s -     20.000  in  61.407844s
common table expressions, limit 300000
                          0.210  (± 0.0%) i/s -     13.000  in  62.433477s
cache_version w/ MD5 checksum, limit 300000
                          0.096  (± 0.0%) i/s -      6.000  in  62.333182s
current cache_version, limit 400000
                          0.298  (± 0.0%) i/s -     17.000  in  60.079039s
common table expressions, limit 400000
                          0.176  (± 0.0%) i/s -     11.000  in  62.762356s
cache_version w/ MD5 checksum, limit 400000
                          0.082  (± 0.0%) i/s -      5.000  in  61.180664s
current cache_version, limit 500000
                          0.237  (± 0.0%) i/s -     14.000  in  60.591729s
common table expressions, limit 500000
                          0.162  (± 0.0%) i/s -     10.000  in  62.121001s
cache_version w/ MD5 checksum, limit 500000
                          0.066  (± 0.0%) i/s -      4.000  in  60.279141s
current cache_version, limit 600000
                          0.228  (± 0.0%) i/s -     14.000  in  62.951441s
common table expressions, limit 600000
                          0.148  (± 0.0%) i/s -      9.000  in  61.489058s
cache_version w/ MD5 checksum, limit 600000
                          0.058  (± 0.0%) i/s -      4.000  in  68.884699s
current cache_version, limit 700000
                          0.191  (± 0.0%) i/s -     12.000  in  63.534652s
common table expressions, limit 700000
                          0.141  (± 0.0%) i/s -      9.000  in  63.893693s
cache_version w/ MD5 checksum, limit 700000
                          0.050  (± 0.0%) i/s -      3.000  in  60.529919s
current cache_version, limit 800000
                          0.178  (± 0.0%) i/s -     11.000  in  63.929976s
common table expressions, limit 800000
                          0.133  (± 0.0%) i/s -      8.000  in  60.379705s
cache_version w/ MD5 checksum, limit 800000
                          0.046  (± 0.0%) i/s -      3.000  in  64.997166s
current cache_version, limit 900000
                          0.191  (± 0.0%) i/s -     12.000  in  64.194715s
common table expressions, limit 900000
                          0.121  (± 0.0%) i/s -      8.000  in  66.573088s
cache_version w/ MD5 checksum, limit 900000
                          0.042  (± 0.0%) i/s -      3.000  in  71.590107s
current cache_version, limit 1000000
                          0.227  (± 0.0%) i/s -     14.000  in  62.149652s
common table expressions, limit 1000000
                          0.119  (± 0.0%) i/s -      8.000  in  67.486037s
cache_version w/ MD5 checksum, limit 1000000
                          0.038  (± 0.0%) i/s -      3.000  in  78.325545s
```

alipman88 commented:
Another means to capture first and last ID would be to use multiple queries, wrapped in a repeatable read transaction, e.g. first `SELECT COUNT(*) AS size, MAX(updated_at) AS time` followed by `SELECT id AS first_id LIMIT 1`, `SELECT id AS last_id LIMIT 1 OFFSET #{size - 1}`. This would provide avoid requiring different SQL code for different RDBMS's - however, setting a nested transaction's isolation level inside an outer transaction results in an error, `ActiveRecord::TransactionIsolationError: cannot set transaction isolation in a nested transaction`. This would occur if `ActiveRecord::Relation#cache_version` was called during an ActiveRecord callback.

So I've ruled that option out, but wanted to note the reasoning.

jonathanhefner commented:
@alipman88  Interesting...

> on large collections, the first and last ID approach takes about 50% longer than the current approach, and the MD5 checksum approach takes about 290% longer than the current approach

Would you mind throwing a few smaller page sizes into the benchmark, like 10, 20, and 50?  Maybe even 200 and 500?  Those sizes should be much more common than e.g. 100000, and it seems like the connection overhead might dwarf the query overhead.  In which case that "50% longer" may not matter as much.

alipman88 commented:
@jonathanhefner, I've added a chart for smaller collections. At or below collection sizes of 1,000, there's no observable difference - any variance in benchmarking results is due to random noise:

https://docs.google.com/spreadsheets/d/1VEH6Rq2lN7nyNCwbVNC5lnOs2TKa1R8Vgh9Vc3_UdcY/edit#gid=1067665996

Playing around with Postgresql's EXPLAIN ANALYZE tool, collection sizes of 50,000 records is around where we start to see different strategies diverge in - and even at that point, it's only a matter of 20 milliseconds or so:

current version:
```
EXPLAIN ANALYZE SELECT
  COUNT(*) AS size,
  MAX(updated_at) AS timestamp
FROM (SELECT updated_at FROM users ORDER BY id LIMIT 50000 OFFSET 1000000) q;

                                                                   QUERY PLAN
-------------------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=18763.56..18763.56 rows=1 width=16) (actual time=246.598..246.598 rows=1 loops=1)
   ->  Limit  (cost=17679.59..18563.56 rows=50000 width=16) (actual time=229.947..243.148 rows=50000 loops=1)
         ->  Index Scan using users_pkey on users  (cost=0.09..35359.09 rows=2000000 width=16) (actual time=0.021..208.578 rows=1050000 loops=1)
 Planning Time: 0.093 ms
 Execution Time: 246.638 ms
(5 rows)
```

first & last ID:
```
EXPLAIN ANALYZE WITH
  q1 AS (SELECT id, updated_at FROM users ORDER BY id LIMIT 50000 OFFSET 1000000),
  q2 AS (SELECT COUNT(*) AS size, MAX(updated_at) AS timestamp FROM q1)
SELECT
  q2.size AS size,
  q2.timestamp AS timestamp,
  (SELECT id FROM q1 LIMIT 1) AS fst_id,
  (SELECT id FROM q1 LIMIT 1 OFFSET (SELECT q2.size FROM q2) - 1) AS lst_id
FROM q2;

                                                                    QUERY PLAN
---------------------------------------------------------------------------------------------------------------------------------------------------
 CTE Scan on q2  (cost=18943.58..18943.59 rows=1 width=32) (actual time=266.407..266.408 rows=1 loops=1)
   CTE q1
     ->  Limit  (cost=17679.59..18563.56 rows=50000 width=16) (actual time=230.042..243.885 rows=50000 loops=1)
           ->  Index Scan using users_pkey on users  (cost=0.09..35359.09 rows=2000000 width=16) (actual time=0.020..209.528 rows=1050000 loops=1)
   CTE q2
     ->  Aggregate  (cost=350.00..350.00 rows=1 width=16) (actual time=259.994..259.994 rows=1 loops=1)
           ->  CTE Scan on q1  (cost=0.00..300.00 rows=50000 width=8) (actual time=230.046..255.724 rows=50000 loops=1)
   InitPlan 3 (returns $2)
     ->  Limit  (cost=0.00..0.01 rows=1 width=8) (actual time=0.002..0.003 rows=1 loops=1)
           ->  CTE Scan on q1 q1_1  (cost=0.00..300.00 rows=50000 width=8) (actual time=0.001..0.001 rows=1 loops=1)
   InitPlan 5 (returns $4)
     ->  Limit  (cost=30.01..30.01 rows=1 width=8) (actual time=6.405..6.405 rows=1 loops=1)
           InitPlan 4 (returns $3)
             ->  CTE Scan on q2 q2_1  (cost=0.00..0.01 rows=1 width=8) (actual time=0.001..0.001 rows=1 loops=1)
           ->  CTE Scan on q1 q1_2  (cost=0.00..300.00 rows=50000 width=8) (actual time=0.001..4.843 rows=50000 loops=1)
 Planning Time: 0.148 ms
 Execution Time: 266.897 ms
```

MD5 checksum:
```
EXPLAIN ANALYZE SELECT
  COUNT(*) AS size,
  MAX(updated_at) AS timestamp,
  SUM(18446744073709551616 * ('x' || LEFT(MD5(id::text), 16))::bit(64)::bigint::decimal(58) +
    ('x' || RIGHT(MD5(id::text), 16))::bit(64)::bigint::decimal(58)) AS checksum
FROM (SELECT id, updated_at FROM users ORDER BY id LIMIT 10000 OFFSET 1000000) q;

                                                                   QUERY PLAN
-------------------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=18021.38..18021.38 rows=1 width=48) (actual time=256.792..256.792 rows=1 loops=1)
   ->  Limit  (cost=17679.59..17856.38 rows=10000 width=16) (actual time=230.085..234.023 rows=10000 loops=1)
         ->  Index Scan using users_pkey on users  (cost=0.09..35359.09 rows=2000000 width=16) (actual time=0.021..201.980 rows=1010000 loops=1)
 Planning Time: 0.109 ms
 Execution Time: 256.847 ms
(5 rows)
```

alipman88 commented:
OK, after some finalizing changes, this pull request is ready for a second review! I've updated the summary at the top and added more comprehensive benchmarks, with charts including linear and logarithmic X-axes.

One spot I'd particularly appreciate a second set of eyes on is in initializing the MySQL @first_id and @last_id variables. As database connections are shared between threads via a connection pool, the same connection must be used to initialize the variables and execute the query. I believe the following code ensures this, but would appreciate verification.

```
connection_pool.with_connection do
  if connection.adapter_name == "Mysql2" && collection.has_limit_or_offset?
    connection.execute("SET @first_id := @last_id := NULL")
  end

  result = connection.select_one(arel, nil)
end
```

alipman88 commented:
Made some marginal improvements to the SQL queries. For MySQL, I moved the `@first_id` variable assignment inside the `COALESCE()` function to avoid unnecessary variable assignment:

```SQL
-- before
SELECT @first_id := COALESCE(@first_id, id)

-- after
SELECT COALESCE(@first_id, @first_id := id)
```

For Postgres, I refactored some raw SQL into Arel and added a conditional to avoid a sub-select that's only necessary for SQLite:

```SQL
-- before
SELECT id AS last_id OFFSET GREATEST((SELECT agg.size FROM agg) - 1, 0);

-- after
SELECT id AS last_id OFFSET GREATEST(agg.size - 1, 0);
```

Benchmarks updated accordingly.

victor-am commented:
I didn't have time to dig into an actual review yet but while scanning through the files I was wondering if this empty `activerecord/db/primary.sqlite3` could be removed 🤔

kaspth commented:
I've only been following along tangentially with what's been going on to help fix the collection cache versions. Can anybody give me a brief update and which PRs should be reviewed?

jonathanhefner commented:
@kaspth #34408 is a good summary and illustration of the issue.  I think this PR is the latest and greatest approach to solving it.  A less complicated alternative approach is #21503, but it retrieves all the IDs and timestamps for a result set, which may be self-defeating (performance-wise) for large result sets.

jonathanhefner commented on activerecord/lib/active_record/integration.rb:
```diff
@@ -161,8 +161,8 @@ def to_param(method_name = nil)
-      def collection_cache_key(collection = all, timestamp_column = :updated_at) # :nodoc:
-        collection.send(:compute_cache_key, timestamp_column)
+      def collection_cache_key(collection = all, id_column: :id, timestamp_column: :updated_at) # :nodoc:
+        collection.send(:compute_cache_key, { id_column: id_column, timestamp_column: timestamp_column })
```
Can we drop the explicit Hash allocation here (i.e. just pass the keyword args)?

alipman88 replied:
Good catch, I've rebased accordingly.

jonathanhefner commented on activerecord/lib/active_record/relation.rb:
```diff
@@ -304,84 +304,151 @@ def many?
+
         if size > 0
-          timestamp = max_by(&timestamp_column)._read_attribute(timestamp_column)
+          timestamp = max_by(&timestamp_column)._read_attribute(timestamp_column).utc.to_s(cache_timestamp_format)
```
I see that the previous code used the timestamp as-is.  Does converting the timestamp to UTC fix another issue?

alipman88 replied:
The previous code did convert to UTC, just on a different line ([line 381 of relation.rb](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/relation.rb#L381)).

Not certain the reasoning behind the UTC conversion. I looked at the [original pull request implementing cache keys](https://github.com/rails/rails/pull/20884), and couldn't find an explanation. Perhaps it avoids the possibility mismatching keys if one's Rails app & database use different time zones.

jonathanhefner commented on activerecord/lib/active_record/relation.rb:
```diff
@@ -304,84 +304,151 @@ def many?
-          subquery_column = "#{subquery_alias}.collection_cache_key_timestamp"
-          arel = query.build_subquery(subquery_alias, select_values % subquery_column)
+          case connection.adapter_name
+          when "PostgreSQL", "SQLite"
```
I'm not familiar with Rails DB-related internals, but I think this kind of branching will need to be abstracted such that every connection adapter can provide its own implementation.  #37944 or #37973 might be useful once merged?

alipman88 replied:
Avoiding this type of branching would be ideal, but I feel the code is so specific to this one use case that abstracting it out to the connection adapter level would make the code harder to follow. Leaving it here for now, but welcome pushback.

I am monitoring those pull requests - once merged, we'll have the opportunity to tidy up some of the Arel.

alipman88 replied:
After thinking about this some more, moving the Postgres/MySQL/SQLite-specific logic to the adapter level is starting to feel like the correct "Rails" way to structure this code.

I've drafted a refactored PR, accordingly: #38171

I'm still not fully certain which direction to go in, so I'm leaving this PR as is (for now). But I'm prepared to move forward with either option.

alipman88 replied:
After chewing on this a little more, I've rebased this PR to move the Postgres/MySQL/SQLite-specific logic into connection adapters.

The previous `ActiveRecord::Relation#compute_cache_version` method is archived here, for reference: https://gist.github.com/alipman88/02f6dd8af2879b597bb9862b9ddc96b8

jonathanhefner commented on activerecord/lib/active_record/relation.rb:
```diff
@@ -304,84 +304,151 @@ def many?
-      else
-        "#{size}"
-      end
+      [size, first_id, last_id, timestamp].reject(&:blank?).join("-")
```
Is there any harm in leaving blank values in the key, i.e. `"#{size}-#{first_id}-#{last_id}-#{timestamp}"`?  That would save two array allocations.

alipman88 replied:
I'd like to avoid making necessary changes to the existing cache key algorithm, per DHH's comment, [here](https://github.com/rails/rails/pull/29092#issuecomment-301503121):

> Furthermore, any change to CacheStore has to be backwards compatible. If
you supply the same values and get a different key, that's no bueno.

But, here's a slightly less succinct solution that avoids necessary array allocations:

```ruby
case
when size.zero?
  "0"
when first_id
  "#{size}-#{first_id}-#{last_id}-#{timestamp}"
else
  "#{size}-#{timestamp}"
end
```

I've rebased accordingly, thanks for flagging.

maximeg commented on activerecord/lib/active_record/integration.rb:
```diff
@@ -161,8 +161,8 @@ def to_param(method_name = nil)
 
-      def collection_cache_key(collection = all, timestamp_column = :updated_at) # :nodoc:
-        collection.send(:compute_cache_key, timestamp_column)
+      def collection_cache_key(collection = all, id_column: :id, timestamp_column: :updated_at) # :nodoc:
```
Default for `id_column` is hardcoded to `:id`, should it be `primary_key`?

alipman88 replied:
Thanks for catching that. It should be consistent throughout, I've refactored accordingly.

maximeg commented on activerecord/lib/active_record/relation.rb:
```diff
@@ -304,83 +304,157 @@ def many?
     # If the collection is loaded, the method will iterate through the records
     # to generate the timestamp, otherwise it will trigger one SQL query like:
     #
     #    SELECT COUNT(*), MAX("products"."updated_at") FROM "products" WHERE (name like '%Cosmic Encounter%')
```
As we are one level up in complexity, adding an example query for both cases (PG/SQLlite vs. MySQL) would be nice for anyone to figure out what will happen (or finding this part of the doc starting from a query from its logs) 😃

alipman88 replied:
I'm a little hesitant to explain the underlying interface in too much detail, but the point about folks looking at a query from their logs and searching for more info seems reasonable. So, I've extended the documentation accordingly.

(Also, while my original solution used window functions, I later determined common table expressions were more efficient - I've corrected that as well.)

alipman88 commented:
Hey @kaspth, you mentioned earlier you might be able to find the time to review this PR. Greatly appreciated if that's still the case. Wanted to check in and hopefully nudge this along before trying to attract reviewers via the Rails core Google group. Thanks!

kaspth commented:
Hey, I do want to get this or something like it in for Rails 6.1, it just keeps falling backwards because it's a tricky issue

aried3r commented:
> So the question is: are there also counter-cases that aren't solved by first and last ID (and count and timestamp)?

We use the [`pg_cache_key`](https://github.com/alekseyl/pg_cache_key) gem because in our case we have a permission system that allows users to be added and removed from access to records.

So what can happen for us, in a collection that is using `.order(updated_at: :desc)` is the following:

1. User A has access to `[1, 2, 3]`, ordered by `updated_at`.
1. User A loses access to `2`.
1. User A gains access to `4` which has an `updated_at` between the ones of `1` and `3`, just like `2`.

These changes happen in a short amount of time so the cache is never invalidated between step 2) and 3). So the resulting collection is `[1, 4, 3]`.

* First `id` stayed the same.
* Last `id` stayed the same.
* `count` stayed the same.
* `maximum(updated_at)` stayed the same.

I still welcome this change as it also solves some of our use cases, but once permissions come into play, things become more tricky and so far, only the MD5 solution worked for us.

Is it a strict requirement that SQLite is the lowest common denominator regarding features? I'm only asking because the MD5 solution is fast enough for us since we use it on paginated results, not on collections with 100.000 records or more where performance might regress and as @jonathanhefner said, it might be a more realistic use case for most users.

Performance might not be great, but it would be interesting to see if a fallback to creating the checksum in Ruby for SQLite is that much worse than the proposed solution.

jonathanhefner commented:
@aried3r That is an interesting example!  Would it be possible, in your case, to summarize a user's permissions into a single value, and make that part of the key passed to a cache block?  Like:

```ruby
cache [@models, md5_of_user_permissions] do
  # ...
end
```

aried3r commented:
> Would it be possible, in your case, to summarize a user's permissions into a single value

Well, the MD5 sum of all IDs in the collection, haha.

On a more serious note: it would be possible in some, not all, cases to have such a value. In cases where it's possible, such a value could exists because of our permission tables. In this case every time a permission is removed, updated, added the `maximum(updated_at)` or last `id` would change.

In the case of "access" ("do you have access to this?"), rather than "permission" ("are you allowed to do this?"), it's a bit harder since in that case the "value" you describe is, in fact, the list of records a user has access to, as described in the scenario I posted previously. That's of course an implementation detail of how we do things.

So far we're happy with `pg_cache_key` as it works the way we expect in this scenario, but should it ever break, become incompatible with, or stay behind, Rails we will need to look into other means. I haven't checked, but it might not play nicely with recyclable cache keys that contain versions.

alipman88 commented:
@aried3r - my thinking was that for many to many relations (e.g. an intermediary permissions table between users and records), the `touch: true` option could be used on the `belongs_to` association to bump the updated_at timestamp:

```
class Permission < ApplicationRecord
  belongs_to :user
  belongs_to :record, touch: true
end
```

If there's any renewed interest in this, I'm happy to resolve the merge conflicts.

aried3r commented:
@alipman88, yes, in these cases it works as intended. We just have, very few, cases where this is not the case. Sometimes records are just added to a `has_many` relation and would end up somewhere between the first and last record. Of course we have ways to bust the cache in this situation, but for others, we heavily rely on the aforementioned gem that uses the `MD5` function.

Again, this PR would solve a lot of simpler cases, but I believe a we-can-detect-whenever-a-collection-is-different solution should be the goal. After all, this issue has persisted for a while now.

It would be interesting to see what requirements would need to be fulfilled to get such a solution. Apparently at least SQLite, MySQL and PostgreSQL should have this functionality, but I don't know about other DB adapters.

alipman88 commented:
@aried3r - using `MD5` in conjunction with `STRING_AGG` (per pg_cache_key) works excellently for Postgres.

For SQLite, it seems like the MD5 extension isn't available by default. I'm running version 3.32.2, which was released in 2020:

```sql
select sqlite_version();
-- 3.32.3

select sha1('xyzzy');
-- Error: no such function: sha1
```

For MySQL, the `GROUP_CONCAT` function (equivalent to Postgres's `STRING_AGG`) has a default max length of 1024 characters. I'm hesitant to modify default settings, but the following pattern seems promising and just about as efficient as my initial first_id/last_id/MAX(updated_at) idea, even for large result sets:

```
# MySQL hash-based fingerprint.
# For each row in the result, append the id and updated_at
# columns to the previous hash and generate a new hash.
SET @hash := '';
SELECT @hash FROM (SELECT @hash := MD5(CONCAT(@hash, updated_at, id)) FROM users LIMIT 1000000) as q limit 1;

+----------------------------------+
| @hash                            |
+----------------------------------+
| e28d00028ba367f6cbead649f4283648 |
+----------------------------------+
1 row in set (1.09 sec)
```

```
# MySQL first_id/last_id/MAX(updated_at) based fingerprints
SET @first := NULL; SET @last := NULL;

SELECT @first, @last, MAX(updated_at) FROM (
  SELECT
    COALESCE(@first, @first := id),
    @last := id,
    updated_at
  FROM users
  ORDER BY id LIMIT 1000000
) AS q
LIMIT 1;

+--------+---------+---------------------+
| @first | @last   | MAX(updated_at)     |
+--------+---------+---------------------+
|      1 | 1000001 | 2021-08-05 12:51:07 |
+--------+---------+---------------------+
1 row in set (0.91 sec)
```

Note my previous caveat for Russian doll caching (in the PR summary): using an MD5 hash may still fail to invalidate the outermost hash key of nested cache fragments. But aside from not covering SQLite, I agree it's a more robust solution.

If there's still enough community interest to develop a solution here, I'm excited to get this over the finish line!

alipman88 commented:
Perhaps the best solution right now is to pluck IDs & updated_at timestamps, and create an MD5 hash on the Ruby side. It won't require any DB adapter-specific code, and any performance impact would be negligible.

aried3r commented:
@alipman88 thanks for your continued work on this! We'd also love to close this topic on our side at work, so perhaps we can bring this to a close. Perhaps @jonathanhefner and @kaspth can look over the proposed solutions (`MD5()` for PostgreSQL/MySQL vs MD5 hash on the Ruby side) so we can move this in the right direction?

alipman88 commented:
Here's an implementation of `ActiveRecord::Relation#compute_cache_version` that handles the MD5 hashing on the Ruby side and is compatible with nearly all cases in `collection_cache_key_test.rb`:

```ruby
def compute_cache_version(timestamp_column)
  collection = eager_loading? ? apply_join_dependency : self
  timestamp_column_name = table[timestamp_column].name
  primary_key_column_name = table[primary_key].name

  unless collection.columns_hash[timestamp_column_name]
    raise ActiveRecord::StatementInvalid, "no such column: #{timestamp_column_name}"
  end

  records_to_pluck = loaded? ? records : collection
  plucked_vals = records_to_pluck.pluck(primary_key_column_name, timestamp_column_name)
  id_hash = ActiveSupport::Digest.hexdigest(plucked_vals.map { |record | record[0] }.join("-"))
  timestamp = plucked_vals.map { |record| record[1] }.max

  if timestamp
    "#{id_hash}-#{timestamp.utc.to_formatted_s(cache_timestamp_format)}"
  else
    id_hash
  end
end
```

After I perform some benchmarking and figure out how to accommodate one remaining test case ("cache_key with a relation having custom select and order"), I'll likely submit a new PR (leaving this one unchanged for reference).

alipman88 commented:
Here's a refresher on the strategies we've discussed, and some benchmarking

**First and last ID:** Using SQL window functions, we can add the first and last ID of a collection to cache key versions. As of MySQL 8, this is a universal solution for all three database adapters. (For MySQL 5, we could implement an alternate approach, or just leave things the way they are for now.) The SQL query would look something like this:

```sql
SELECT
  COUNT(*) AS size,
  MAX(first_id) AS first_id,
  MAX(last_id) AS last_id,
  MAX(updated_at) AS timestamp
FROM (
  SELECT
    FIRST_VALUE(id) OVER() AS first_id,
    LAST_VALUE(id) OVER() AS last_id,
    updated_at
  FROM (SELECT * FROM widgets ORDER BY name LIMIT 100) cck_subquery_1
) cck_subquery_2
```

**MD5 hash:** We could pluck IDs & updated at timestamps, generate an MD5 hash of IDs ruby-side. (Sqlite doesn't ship with a built-in MD5 function, limiting a universal SQL-side solution.) This would effectively increase the time complexity of generating a cache key from constant to linear. However, it's more robust in certain cases involving HABTM associations, and IMO somewhat simpler. The Ruby would look something like this:

```ruby
def compute_cache_version_md5(timestamp_column, id_column)
  collection = eager_loading? ? apply_join_dependency : self
  id_column_name = table[id_column].name
  timestamp_column_name = table[timestamp_column].name
  subquery = collection.select(id_column_name, timestamp_column_name)
  query = collection.klass.select(id_column_name, timestamp_column_name).from(subquery)
  records_to_pluck = loaded? ? records : query
  plucked_records = records_to_pluck.pluck(id_column_name, timestamp_column_name)
  digest = ActiveSupport::Digest.hash_digest_class.new
  id_hash = plucked_records.map(&:first).map(&:to_s).reduce(digest) { |md5, id| md5 << id }.hexdigest
  timestamp = plucked_records.map(&:second).max

  if timestamp
    "#{id_hash}-#{timestamp.utc.to_formatted_s(cache_timestamp_format)}"
  else
    id_hash
  end
```

Here's a graph showing the time tradeoffs between the current strategy, a first & last ID strategy (using SQL window functions), and a Ruby-side MD5 strategy:

![Screen Shot 2022-02-13 at 1 46 50 PM](https://user-images.githubusercontent.com/3922511/153771126-99618b68-1afa-4d2a-86b7-2dfb382d2ef6.png)

Here's the benchmarking script used to generate the above graph, raw results: https://gist.github.com/alipman88/9242d30eb55e923545dde145db05324b

An increase of a few milliseconds to compute versioned cache keys for really large collections doesn't feel like an intolerable cost, but I'm still considering what the best/most universal solution might be.

marckohlbrugge commented:
I know this is an old PR, but the bug is still present in Rails 7.1.3

Is there any appetite in merging this into `main` ?
